### PR TITLE
Bring acceptance tests to the next level 🚀

### PR DESCRIPTION
**Pull request by EmilienM**
# Bring acceptance tests to the next level

This PR has 7 commits which should bring the acceptance jobs in a better shape than they are today:
* Move acceptance tests to a generic script
* zuul: split the jobs to spread the load
* zuul: optimize how jobs are triggered
* zuul: add variants for stable jobs
* zuul/stable: fix Ubuntu versions
* Pre-fetch the Octavia Amphora image
* Disable testing for neutron trunk & octavia loadbalancers

## Move acceptance tests to a generic script

This patch will avoid the duplication of gophercloud-acceptance script
by adding job variants, where we set variables of what we want to deploy
and it'll use the same workflow/scripts.

This removes the specific playbook for running Ironic tests and creates
these variables:

* `acceptance_tests`: List of acceptance tests (which can be discovered
  via:
```
find acceptance/openstack -name '*_test.go' -exec dirname {} \; | sort -n | uniq -c | awk '{print $2}'
```

* `devstack_services`: List of devstack services to activate.
* `devstack_override_enabled_services`: List of services to activate in
  Devstack, specific to the framework itself.
* `devstack_projects`: List of projects that will have to be cloned by
  devstack to be installed.

This moves both gophercloud-acceptance-test and
gophercloud-acceptance-test-ironic to use this new layout; providing the
same configuration as before.

This will reduce maintenance cost and simplify when we want to add other
variants (will come later).

## zuul: split the jobs to spread the load

* `gophercloud-acceptance-test` becomes
   `gophercloud-acceptance-test-compute` which will test Nova, Zun,
    Keystone, Barbican, Heat and Placement.
* `gophercloud-acceptance-test-networking` which will test Neutron,
    Designate and Octavia.
* `gophercloud-acceptance-test-storage` which will test Cinder, Glance,
    Swift and Manila.

This will hopefully helps to reduce the timeouts when running everyone within a single instance and spread the load across 3 jobs instead of 1.

## zuul: optimize how jobs are triggered

* Run `gophercloud-acceptance-test-compute` all the times, this is our
  "reference" job for OpenStack; testing the minimal set of features.
  Also from now, we won't run this job if documentation is the only
  change in a PR.
* Run the networking & storage & ironic variants only when it's needed, to avoid
  wasting CI resources.

## zuul: add variants for stable jobs

Add the new variants for stable OpenStack platforms.
Note: we had to keep the original job, as "legacy", so we maintain the
same testing matrix.

## zuul/stable: fix Ubuntu versions

Fix the Ubuntu versions for the stable OpenStack jobs.
Note: for each version, we had a look at:

https://github.com/openstack/devstack/blob/stable/<version-name>/stack.sh

## Pre-fetch the Octavia Amphora image

This will avoid issues that we currently have with DIB, until we figure
it out correctly. To figure it out correctly, we need to rework the
roles that we run for Devstack job to properly collect the logs.
This will be done in the future but not in this first effort.
For now, we'll download the 500MB+ image from RDO servers, which should
be faster and more reliable than building the image locally.

##  Disable testing for neutron trunk & octavia loadbalancers

They currently fail, we need to figure out why:
* All LBs end up being in ERROR state.
* Trunks are not usable: `Unable to list trunks: Resource not found`.

Issue #2251

### CODE REVIEW

**Comment by EmilienM** [comment]
recheck-ussuri

**Comment by EmilienM** [comment]
recheck stable/ussuri

**Comment by EmilienM** [comment]
recheck stable/ussuri

**Comment by EmilienM** [comment]
recheck stable/ussuri

**Comment by EmilienM** [comment]
One job failed because a crontab decided to run unattended upgrade on the host, while devstack was trying to install packages :-(
```
root       89917   89901   89917  0.0  0.0   320  1644 /bin/sh /usr/lib/apt/apt.systemd.daily lock_is_held update
root       90375   89917   90375 98.6  1.0 83516 84320 /usr/bin/python3 /usr/bin/unattended-upgrade --download-only
```

**Comment by EmilienM** [comment]
recheck

**Comment by EmilienM** [comment]
recheck

**Comment by EmilienM** [comment]
recheck

**Comment by EmilienM** [comment]
recheck

**Comment by EmilienM** [comment]
recheck

**Comment by EmilienM** [comment]
recheck

**Comment by EmilienM** [comment]
@jtopjian the octavia test fail and neutron trunks as well. See commit message on https://github.com/gophercloud/gophercloud/pull/2252/commits/6c22124b9131b37f57c2061a5f316d9f869d143c
Before I dig into it, do you have any idea what can go wrong?

**Comment by EmilienM** [comment]
recheck
to see if results are consistent

**Comment by jtopjian** [comment]
@EmilienM I'm not able to see anything obvious. It looks like the screen logs stopped logging at 14:38 and the Gophercloud tests begin at 14:39, so we're not able to look at the service-side logs to see what happened.

If you set `OS_DEBUG=1` as an environment variable, Gophercloud will log the full HTTP conversation. This might help provide some additional information as to why the tests are failing.

**Comment by EmilienM** [comment]
recheck
heat is flaking and logs don't help...

**Comment by EmilienM** [comment]
> @EmilienM I'm not able to see anything obvious. It looks like the screen logs stopped logging at 14:38 and the Gophercloud tests begin at 14:39, so we're not able to look at the service-side logs to see what happened.
> 
> If you set `OS_DEBUG=1` as an environment variable, Gophercloud will log the full HTTP conversation. This might help provide some additional information as to why the tests are failing.

I'm more interested by OpenStack logs but somehow they don't show up very well: https://logs.openlabtesting.org/logs/52/2252/e92ab4a8be101d0b61ba22d365883679e3d41a37/check/gophercloud-acceptance-test-compute/5e225a1/logs/screen-h-eng.txt
this file is useless for debug :/

**Comment by jtopjian** [comment]
@EmilienM Yeah, exactly. I'm confused about why there's a lack of service-side logs for each of the tests being run. Since DEBUG is enabled on the service-side, there should be log entries for each Gophercloud HTTP request being made and the subsequent internal tasks being done.

I'm not sure why things aren't logging, though - both Zuul and how OpenLab is configured is a mystery to me.

The `OS_DEBUG=1` setting for Gophercloud is something I can confidently suggest to provide additional logging, but only from the client point of view.

**Comment by EmilienM** [comment]
recheck

**Comment by EmilienM** [comment]
@jtopjian what do you think about this effort? I know this is a big PR, I can split it out but this is what I had in mind to "stabilize" acceptance jobs in the short term, until we decide what we do in the future; and still being able to scale the jobs (adding more services, without burning the ex-single VM).

**Comment by jtopjian** [comment]
@EmilienM This looks good and makes sense to me. Nice work 🙂

**Comment by EmilienM** [comment]
@mandre @jtopjian for review, when you have time.
Martin will probably have some feedback but won't have the right to merge it. Joe, if you are good with it, just +1 for now and I'll wait for Martin's review.

Thanks

**Review by jtopjian — APPROVED** [review]
This LGTM. 

It looks like there are two core themes with this PR:

1. Skipping tests that are not running correctly in OpenLab.

   In this situation, the issue is with OpenLab and not the tests themselves, so I'm good with configuring them to be skipped.

2. Refactoring Zuul

   These changes make sense to me.

**Review by mandre — COMMENTED** [review]
Cool stuff - this is a good improvement over the existing. I left a couple of comments but roughly my questions can be summarized as:
- what versions of OpenStack do we want to support? The jobs tests will need to reflect that, and later we'll want to add jobs for OpenStack releases > Ussuri too.
- why do we disable tests? We need a very good reason for disabling tests, and if we do, we want to keep track of it so we don't forget to re-enable them. Otherwise if the tests are no longer valid, just delete them.

**Comment by EmilienM** [comment]
> what versions of OpenStack do we want to support? The jobs tests will need to reflect that,

We support all versions for now (since Newton I believe). This will probably evolve in the future.

> and later we'll want to add jobs for OpenStack releases > Ussuri too

Yes, I already thought about it and prepared zuul jobs: https://github.com/theopenlab/openlab-zuul-jobs/commit/c350c09e1593672d6839b036250e6d5563287b3d

File: .zuul.yaml [hunk 1]
```diff
@@ -7,43 +7,419 @@
...
+      - ^LICENSE$
+
+- job:
+    name: gophercloud-acceptance-test-compute
+    parent: gophercloud-acceptance-test-base
+    description: |
+      Run gophercloud compute acceptance test on master branch. This runs when any file is patched
+      except if it's doc.
+    vars:
+      # the list of all available tests can generated by:
```
[the remark this answers is not in the document]
Comment (EmilienM) [reply .zuul.yaml:29]: good idea. I created an issue for it so we don't forget about it: https://github.com/gophercloud/gophercloud/issues/2263

File: .zuul.yaml [hunk 2]
```diff
@@ -7,43 +7,419 @@
...
+      - ^.*dns.*$
+      - ^.*loadbalancer.*$
+      - ^.*networking.*$
+    vars:
+      prefetch_amphora: true
+      devstack_env:
+        OCTAVIA_AMP_IMAGE_FILE: /opt/octavia-amphora/amphora-x64-haproxy.qcow2
+      # the list of all available tests can generated by:
+      # find acceptance/openstack -name '*_test.go' -exec dirname {} \; | sort -n | uniq -c | awk '{print $2}'
+      acceptance_tests:
```
[the remark this answers is not in the document]
Comment (EmilienM) [reply .zuul.yaml:60]: very cool, thanks. I'll give this a try!

File: acceptance/openstack/loadbalancer/v2/loadbalancers_test.go [hunk 3]
```diff
@@ -45,6 +45,7 @@ func TestLoadbalancersListByTags(t *testing.T) {
 	clients.SkipRelease(t, "stable/pike")
 	clients.SkipRelease(t, "stable/queens")
 	clients.SkipRelease(t, "stable/rocky")
+	t.Skip("Currently failing in OpenLab")
```
[the remark this answers is not in the document]
Comment (EmilienM) [reply acceptance/openstack/loadbalancer/v2/loadbalancers_test.go:48]: yeah this is something we should address later and have issues files for each of them. For now I have no idea why they fail and i'm trying to fix one problem at a time.

File: .zuul.yaml [hunk 4]
```diff
@@ -7,43 +7,419 @@
...
+    name: gophercloud-acceptance-test-storage-queens
+    parent: gophercloud-acceptance-test-storage
+    description: |
+      Run gophercloud storage acceptance test on queens branch
+    nodeset: ubuntu-xenial
+    vars:
+      global_env:
+        OS_BRANCH: stable/queens
+
+# NOTE: A Pike-based devstack environment is currently
```
[the remark this answers is not in the document]
Comment (EmilienM) [reply .zuul.yaml:271]: I agree a lot with these statements, however we should address this in a future PR and communicate with our community about that change.

Thanks

File: script/acceptancetest [hunk 5]
```diff
@@ -7,97 +7,15 @@ source `dirname $0`/stackenv
...
-# No suitable endpoint could be found in the service catalog
-# acceptance/openstack/workflow/v2
-)
+if [[ -z $ACCEPTANCE_TESTS ]]; then
+    echo "No acceptance tests to run"
+    exit 0
+fi
 
-for acceptance_test in ${ACCEPTANCE_TESTS[@]}; do
+for acceptance_test in "${ACCEPTANCE_TESTS[@]}"; do
```
Comment (mandre) [inline script/acceptancetest:16]: Remove the quotes, it causes the loop to fail:
```
2021-11-09 16:07:40.477103 | ubuntu-focal | + ./script/acceptancetest:main:16          :   for acceptance_test in "${ACCEPTANCE_TESTS[@]}"
2021-11-09 16:07:40.484427 | ubuntu-focal | + ./script/acceptancetest:main:17          :   go test -v -timeout 60m -tags 'fixtures acceptance' ./acceptance/openstack/blockstorage acceptance/openstack/blockstorage/extensions acceptance/openstack/blockstorage/noauth acceptance/openstack/blockstorage/v1 acceptance/openstack/blockstorage/v2 acceptance/openstack/blockstorage/v3 acceptance/openstack/imageservice/v2 acceptance/openstack/objectstorage/v1 acceptance/openstack/sharedfilesystems/v2 acceptance/openstack/sharedfilesystems/v2/messages
2021-11-09 16:07:40.538842 | ubuntu-focal | package acceptance/openstack/blockstorage/extensions is not in GOROOT (/usr/local/go/src/acceptance/openstack/blockstorage/extensions)
2021-11-09 16:07:40.539032 | ubuntu-focal | package acceptance/openstack/blockstorage/noauth is not in GOROOT (/usr/local/go/src/acceptance/openstack/blockstorage/noauth)
2021-11-09 16:07:40.539145 | ubuntu-focal | package acceptance/openstack/blockstorage/v1 is not in GOROOT (/usr/local/go/src/acceptance/openstack/blockstorage/v1)
2021-11-09 16:07:40.539269 | ubuntu-focal | package acceptance/openstack/blockstorage/v2 is not in GOROOT (/usr/local/go/src/acceptance/openstack/blockstorage/v2)
2021-11-09 16:07:40.539386 | ubuntu-focal | package acceptance/openstack/blockstorage/v3 is not in GOROOT (/usr/local/go/src/acceptance/openstack/blockstorage/v3)
2021-11-09 16:07:40.539510 | ubuntu-focal | package acceptance/openstack/imageservice/v2 is not in GOROOT (/usr/local/go/src/acceptance/openstack/imageservice/v2)
2021-11-09 16:07:40.539621 | ubuntu-focal | package acceptance/openstack/objectstorage/v1 is not in GOROOT (/usr/local/go/src/acceptance/openstack/objectstorage/v1)
2021-11-09 16:07:40.539738 | ubuntu-focal | package acceptance/openstack/sharedfilesystems/v2 is not in GOROOT (/usr/local/go/src/acceptance/openstack/sharedfilesystems/v2)
2021-11-09 16:07:40.540412 | ubuntu-focal | package acceptance/openstack/sharedfilesystems/v2/messages is not in GOROOT (/usr/local/go/src/acceptance/openstack/sharedfilesystems/v2/messages)
```

https://logs.openlabtesting.org/logs/52/2252/6ceef7690d7fef7fddfb83756c8446d2c63a5139/check/gophercloud-acceptance-test-storage/df72cf6/job-output.txt.gz

Comment (EmilienM) [reply script/acceptancetest]: @mandre this patch causes too much troubles for now, I want to do this, but in a next step, and make it clean.
For now, I would like to restore a good shape for CI and this PR is really too big.. I don't want to make it bigger.

Thanks

Comment (mandre) [reply script/acceptancetest]: Fair enough, let's merge this PR.
I created https://github.com/gophercloud/gophercloud/issues/2267 to keep track of the skipped tests.

**Comment by mandre** [comment]
This is step 1 towards getting more value out of the acceptance tests, get the tests to green and decrease noise. Let's improve further in follow-up PRs.

LGTM

**Pull request merged**